### ISSUE TITLE
shared geostory columns links sometimes it scrolls to related section instead

### ISSUE THREAD

**Issue by MV88** [geosolutions-it/MapStore2#6830]
## Description

see https://github.com/geosolutions-it/webmapper-halliburton/issues/1745
there is a wrong order of actions that makes the scroll happen to the non desired location, i.e. column

## How to reproduce

-  open this https://dev-mapstore.geosolutionsgroup.com/mapstore/#/geostory/shared/11555/section/7eca906c-5dae-431c-b313-80721d4af301/column/08b82a2d-057c-4a22-991d-92911b115d86

*Expected Result*

you should see the url is maintained in the address bar i.e.
https://dev-mapstore.geosolutionsgroup.com/mapstore/#/geostory/shared/11555/section/7eca906c-5dae-431c-b313-80721d4af301/column/08b82a2d-057c-4a22-991d-92911b115d86

![image](https://user-images.githubusercontent.com/11991428/117163522-04c2c280-adc4-11eb-9226-20ded71db933.png)

*Current Result*

url is restricted to section
https://dev-mapstore.geosolutionsgroup.com/mapstore/#/geostory/shared/11555/section/7eca906c-5dae-431c-b313-80721d4af301

![image](https://user-images.githubusercontent.com/11991428/117163579-14420b80-adc4-11eb-8097-fe9bde9255e2.png)

- [x] Not browser related

<details><summary> <b>Browser info</b> </summary>

(use this site: <a href="https://www.whatsmybrowser.org/">https://www.whatsmybrowser.org/</a> for non expert users)

| Browser Affected | Version |
|---|---|
|Internet Explorer| |
|Edge| |
|Chrome| |
|Firefox| |
|Safari| |
</details>

**Comment by offtherailz** [comment]
From an investigation an a test, disabling URL update when applying the scroll event looks to do the job. I gave 500ms buffer before and after the scroll. 
In this interval the URL sync is  disabled by semaphore, so the UPDATE_CURRENT_PAGE events will be ignored

This is my test code, replacing `urlUpdateOnScroll` and `scrollOnLoad` epics with the following (adding `semaphore` utility). 
Of course the action is hard coded in this case, while it have to be 
```javascript
const semaphore = (sem$, start = true, condition = (c) => c) => (stream$) =>
    stream$
        .withLatestFrom(sem$.startWith(start))
        .filter(([, s]) => condition(s))
        .map(([e]) => e);

/**
 * Handle the url updates on currentPage change
 * @param {Observable} action$ stream of actions
 * @param {object} store redux store
 */
export const urlUpdateOnScroll = (action$, {getState}) =>
    action$.ofType(UPDATE_CURRENT_PAGE)
        .let(semaphore(
            action$.ofType("GEOSTORY:SCROLLING")
                .map(a => !a.value)
        ))
        .debounceTime(50) // little delay if too many UPDATE_CURRENT_PAGE actions come
        .switchMap(({sectionId, columnId}) => {
            if (
                modeSelector(getState()) !== 'edit' && (!!columnId || !!sectionId)
                && updateUrlOnScrollSelector(getState())
            ) {
                const storyId = get(getState(), 'geostory.resource.id');
                const newUrl = parseHashUrlScrollUpdate(window.location.href, window?.location?.hash, storyId, sectionId, columnId);
                if (newUrl) {
                    history.pushState(null, '', newUrl);
                }
        }
            return Observable.empty();
        });

/**
 * When story is loaded, scrolls down to the element from the URL
 * @param {Observable} action$ stream of actions
 */
export const scrollOnLoad = (action$) =>
    action$.ofType(SET_CURRENT_STORY)
        .switchMap(() => {
            const storyIds = window?.location?.hash?.split('/');
            return Observable.of(storyIds)
                .delay(500)
                .do(() => {
                    if (window?.location?.hash?.includes('shared')) {
                            scrollToContent(storyIds[7] || storyIds[5], {block: "start", behavior: "auto"});
                        } else if (storyIds.length > 5) {
                            scrollToContent(storyIds[6], {block: "start", behavior: "auto"});
                        } else if (storyIds.length === 5) {
                            scrollToContent(storyIds[4], {block: 'start', behavior: "auto"});
                        }
                    }
                )
                .ignoreElements()
                .startWith({type: "GEOSTORY:SCROLLING", value: true})
                .concat(Observable.of({type: "GEOSTORY:SCROLLING", value: false}).delay(500));
    });
```

**Comment by ElenaGallo** [comment]
@MV88 this issue is still present, only after selecting another Immersive Content then the url of the first Immersive Content is displayed correctly, see the gif below:

![share_story](https://user-images.githubusercontent.com/56537133/117994459-da7f8080-b340-11eb-8619-97bacf00f86e.gif)

**Comment by offtherailz** [comment]
I think this is not a problem, because sharing the 2 URL should give the same result, isn't it?

**Comment by ElenaGallo** [comment]
@offtherailz the two results are different because the first URL is: 

> https://dev-mapstore.geosolutionsgroup.com/mapstore/#/geostory/shared/11555/section/7eca906c-5dae-431c-b313-80721d4af301

and the second URL is: 

> https://dev-mapstore.geosolutionsgroup.com/mapstore/#/geostory/shared/11555/section/7eca906c-5dae-431c-b313-80721d4af301/column/7ee2c8a2-e960-45aa-b008-f75589606d6b

**Comment by offtherailz** [comment]
Yes, I mean that, whatever is the URL, I will land to the same position of the page (same scroll).

In fact this happens because the section URL may or not may contain columns (depending on column type, for instance tiles).
So some scroll position may have multiple URLs, referring the same place. 
In particular the first column of the immersive section has the url like /section/<sectionId> and /section/<sectionId>/column/<first-column-id> but they bring to the same point in page. Scrolling up and down events update columns and section while you scroll, so the direction may influence the event sequence, but the important thing is that they point always to the current section. 

Anyway I noticed that sometimes, randomly the URL is replaced by the application in this way:

https://dev-mapstore.geosolutionsgroup.com/mapstore/#/geostory/shared/11555/section/title_section_id1/column/b9b758ba-bf93-48f5-9bcd-92292e7d9858

That is wrong (`title_section_id1` is not correct), even if it works anyway. 
Other times I have this URL and the scroll position is not correct:
https://dev-mapstore.geosolutionsgroup.com/mapstore/#/geostory/shared/11555/section/title_section_id1

This is effectively a bug that have to be fixed. 

For devs (@MV88 ) I didn't noticed an issue like this during development.

I think this is due to the scroll event on page load (the ID is referred to the first section) that has a race condition with the one coming from the URL.

**Comment by tdipisa** [comment]
@offtherailz should we open a separate issue for this?

>Anyway I noticed that sometimes, randomly the URL is replaced by the application in this way:

>https://dev-mapstore.geosolutionsgroup.com/mapstore/#/geostory/shared/11555/section/title_section_id1/column/b9b758ba-bf93-48f5-9bcd-92292e7d9858

>That is wrong (title_section_id1 is not correct), even if it works anyway.
Other times I have this URL and the scroll position is not correct:
https://dev-mapstore.geosolutionsgroup.com/mapstore/#/geostory/shared/11555/section/title_section_id1

>This is effectively a bug that have to be fixed.

**Comment by offtherailz** [comment]
Maybe is a good idea, because this is not the same problem in the title. Anyway this also may influence the project labeled. So I suggest to verify if the problem happens also in that environment @MV88

**Comment by offtherailz** [comment]
I hopened #6970 for the new issue. This is closed IMHO. 
The issue notified by @ElenaGallo is a consistent behavior, anyway.

The positon at the start of the immersive section has 2 possible URLs. the application uses one of them depending on the event sequence
I think is tolarable.
